gobblin-tunnel/src/main/java/gobblin/tunnel/Tunnel.java

```diff
@@ -0,0 +1,178 @@
+
+  private ServerSocketChannel server;
+  private Thread thread;
+  private final Config config;
+
+  private Tunnel(String remoteHost, int remotePort, String proxyHost, int proxyPort) {
+    config = new Config(remoteHost, remotePort, proxyHost, proxyPort);
+  }
+
+  private Tunnel open() throws IOException {
+    try {
+      server = ServerSocketChannel.open().bind(null);
+      startTunnelThread();
+
+      return this;
+    } catch (IOException ioe) {
+      LOG.error("Failed to open the tunnel", ioe);
+      throw ioe;
+    }
+  }
+
+  public int getPort() throws IOException {
+    SocketAddress localAddress = null;
+
+    try {
+      if (server != null && server.isOpen()) {
+        localAddress = server.getLocalAddress();
+      }
+      if (localAddress instanceof InetSocketAddress) {
+        return ((InetSocketAddress) localAddress).getPort();
+      }
+    } catch (IOException e) {
+      LOG.error("Failed to get tunnel port", e);
+      throw e;
+    }
+    return NON_EXISTENT_PORT;
+  }
+
+  private void startTunnelThread() {
+    thread = new Thread(new Dispatcher(), "Tunnel Listener");
+    thread.start();
+  }
+
+  public boolean isTunnelThreadAlive() {
+    return (thread != null && thread.isAlive());
+  }
+
+  private class Dispatcher implements Runnable {
+
+    private Selector selector;
+
+    public Dispatcher() {
+      try {
+        selector = Selector.open();
+      } catch (IOException e) {
+        LOG.error("Could not open selector",e);
+      }
+    }
+
+    @Override
+    public void run() {
+      try {
+        new AcceptHandler(server, selector, config);
+
+        while (!Thread.interrupted()) {
+
+          selector.select();
+          Set<SelectionKey> selectionKeys = selector.selectedKeys();
+
+          for (SelectionKey selectionKey : selectionKeys) {
+            dispatch(selectionKey);
+          }
+          selectionKeys.clear();
+        }
+      } catch (IOException ioe) {
+        LOG.error("Unhandled exception.  Tunnel will close", ioe);
+      }
+
+      LOG.info("Closing tunnel");
+    }
+
+    private void dispatch(SelectionKey selectionKey)
+        throws IOException {
+      Callable<?> attachment = (Callable<?>) selectionKey.attachment();
+
+      try {
+        attachment.call();
+      } catch (Exception e) {
+        LOG.warn("exception handling event on {}",selectionKey.channel(), e);
+      }
+    }
+  }
+
+  public void close() {
+    try {
+      thread.interrupt();
+      thread.join();
+    } catch (InterruptedException e) {
+      throw new RuntimeException(e);
+    } finally {
+      try {
+        server.close();
+        LOG.info("Closed tunnel.");
+      } catch (IOException ioe) {
+        LOG.warn("Failed to shutdown tunnel", ioe);
+      }
+    }
+  }
+
+  public static Tunnel build(String remoteHost, int remotePort, String proxyHost, int proxyPort)
+      throws IOException {
+    return new Tunnel(remoteHost, remotePort, proxyHost, proxyPort).open();
```

Comment:
It's not really URLs, but different hostnames. A Tunnel is currently pre-configured to always connect to the same hostname. But the Tunnel does support multiple concurrent connections to the same target host. So you could potentially use the same Tunnel instance for different URLs on the same hostname, depending on what the client requires and protocol supports.

If you want to connect to different hostnames, yes, you would need multiple instances of the Tunnel (and correspondingly the internal ServerSocketChannel). This is not currently a problem, because a) that would imply multiple JdbcProvider instances would be created, each with its own Tunnel instance and b) currently for JdbcExtractors, Gobblin does not appear to support multiple source.conn.* config settings per extractor anyway.